src/experimental/information_estimates.cpp

```diff
@@ -0,0 +1,383 @@
+        {
+            karto::Vector2<kt_double> tf_laser{laser_readings[i].GetX() - m_low_x, laser_readings[i].GetY() - m_low_y};
+
+            // Laser final cell
+            karto::Vector2<int> beam_grid = utils::grid_operations::getGridPosition(tf_laser, m_cell_resol);
+
+            // Visited cells by this laser beam
+            std::vector<karto::Vector2<int>> cells = utils::grid_operations::rayCasting(robot_grid, beam_grid);
+
+            for (auto & cell : cells)
+            {
+                // Inidividual cell limits
+                kt_double limit_x = cell.GetX() * m_cell_resol;
+                kt_double limit_y = cell.GetY() * m_cell_resol;
+
+                std::pair<std::vector<kt_double>, std::vector<kt_double>> intersections = utils::grid_operations::computeLineBoxIntersection(
+                    robot_pose_tf.GetPosition(), tf_laser, robot_grid, beam_grid, limit_x, limit_y, m_cell_resol);
+
+                if (intersections.first.size() == 0)
+                    continue;
+
+                // Enter (d1) and Exit (d2) distances
+                std::vector<kt_double> distances;
+                for (int k = 0; k < intersections.first.size(); ++k)
+                {
+                    // From robot position to intersection points
+                    karto::Vector2<kt_double> intersection{intersections.first[k], intersections.second[k]};
+                    kt_double distance = robot_pose_tf.GetPosition().Distance(intersection);
+                    distances.push_back(distance);
+                }
+
+                // Measurement outcomes vector {Pfree, Pocc, Pun}
+                std::vector<kt_double> probabilities {
+                    calculateScanMassProbabilityBetween(distances[1], m_max_sensor_range),
+                    calculateScanMassProbabilityBetween(distances[0], distances[1]),
+                    calculateScanMassProbabilityBetween(0.0f, distances[0])
+                };
+
+                // Appending new measurement outcomes for the current cell
+                appendCellProbabilities(probabilities, cell);
+
+                // Get all the measurement outcomes for the current cell
+                std::vector<std::vector<kt_double>> cell_prob = retrieveCellProbabilities(cell);
+
+                // Compute all the possible combinations for the current cell - algorithm 1
+                std::unordered_map<map_tuple, kt_double, utils::tuple_hash::HashTuple> meas_out_prob = computeMeasurementOutcomesHistogram(cell_prob);
+
+                kt_double cell_mutual_inf = 0.0f;
+                for (auto& pair : meas_out_prob)
+                {
+                    cell_mutual_inf +=  pair.second * utils::probability_operations::calculateMeasurementOutcomeEntropy(pair.first);
+                }
+
+                // Mutual information of cell x, y given a set of measurements
+                updateCellMutualInformation(1.0 - cell_mutual_inf, cell);
```

Comment:
@kmilo7204 perf: random access to a matrix in row major (or column major) linear storage will always be slower than sequential access matching storage order (ie. row by row if row major or column by column if column major).

Reply:
See proposal for reversing the iteration order. That would make this issue go away.